lib/service.js

```diff
@@ -0,0 +1,166 @@
+          cc: clean_options.ccAddresses,
+          to: clean_options.toAddresses,
+          html: clean_options.html_body,
+          text: clean_options.txt_body,
+          subject: clean_options.subject,
+          from: clean_options.senderEmailAddress,
+          'h:Reply-To': clean_options.replyToAddress
+        };
+      },
+      send: function (params, callback) {
+        var mg = mailgun.client({
+          username: 'api',
+          key: process.env.MAILGUN_API_KEY
+        });
+
+        return mg.messages.create(process.env.MAILGUN_SENDING_DOMAIN, params)
+          .then(function (message) {
+            return callback(null, message);
+          })
+          .catch(function (err) {
+            return callback(err);
+          });
+      }
+    }
+  },
+
+  ses: {
+    required_env: ['AWS_REGION', 'AWS_ACCESS_KEY_ID', 'AWS_SECRET_ACCESS_KEY'],
+    export: {
+      name: 'ses',
+      parameterize: function (clean_options) {
+        /* eslint-disable max-len */
+        // see: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/SES.html#sendEmail-property
+        return {
+          Destination: {
+            /* required */
+            BccAddresses: clean_options.bccAddresses,
+            CcAddresses: clean_options.ccAddresses,
+            ToAddresses: clean_options.toAddresses
+          },
+          Message: {
+            /* required */
+            Body: {
+              /* required */
+              Html: {
+                Data: clean_options.html_body,
+                /* required */
+                Charset: clean_options.htmlCharset || clean_options.defaultCharset
+              },
+              Text: {
+                Data: clean_options.txt_body,
+                /* required */
+                Charset: clean_options.textCharset || clean_options.defaultCharset
+              }
+            },
+            Subject: {
+              /* required */
+              Data: clean_options.subject,
+              /* required */
+              Charset: clean_options.subjectCharset || clean_options.defaultCharset
+            }
+          },
+          Source: clean_options.senderEmailAddress, /* required */
+          ReplyToAddresses: [
+            clean_options.replyToAddress
+          ]
+        };
+      },
+      send: function (params, callback) {
+        var ses = new AWS.SES({ region: process.env.AWS_REGION });
+
+        return ses.sendEmail(params, callback);
+      }
+    }
+  }
+
+};
+
+function envVarExists (envVar) {
+  return process.env.hasOwnProperty(envVar) && process.env[envVar];
+}
+
+/**
+ * Determines the sending service that the current request to send will use
+ * @param {String} specific_service - name of sending service to use;
+ * properly assuming set configuration, overrides default
+ * @returns {Object} interface for parameterizing and sending email send requests
+ * with the determined service
+ */
+function determine_service (specific_service) {
```

Comment:
I have a feeling there is maybe a nicer way to do this, having this as part of the library code and referencing another `CURRENT_SERVICE` env var feels a little off, let me know your thoughts @nelsonic @zemccartney?

Reply:
@samhstn totally agree, I was never all that happy with how I did things here. 

I came up with a new approach to this in the stuff I just pushed up. 

I still have the library using another env var, now renamed to `SENDEMAIL_SERVICE` to hopefully make the var's name more descriptive/understandable, but only if the user happens to have multiple services configured at once.  In that case, we throw an error saying the user's env config is ambiguous, directing them to use that variable to tell the library which sending service to use. 

In short, I tried to clarify and minimize the purpose of that env var by moving it out of the main path of the `determine_service` path. 

I think I was solving for the, to be honest, nonsensical defaulting behavior I had in place to handle this case before  (essentially, unless the user had set `CURRENT_SERVICE`, the library would use the last available service, in alphabetical order....no good 😆)

Does that get to root of this issue? Or at least improve things a bit?

My bad if not! I just jumped to this approach to try working through another thought. 
I kept `determine_service` for now because I figure we need to handle the case — I hope a seriously unlikely one, but I guess still possible — that a user could configure multiple services at once, like our tests. If that's not important, I bet we can simplify this part of the library a good deal (though not sure of the clearest route for that aside from continuing to use an env var...would have to think about that)

Reply:
@samhstn yo! any thoughts here?

sorry, no rush at all. I don't mean to bug, more chiming back so wrapping this up doesn't completely fall off my radar (again :) ). Let me know if there's anything I can do to help move this forward. Thanks for all the help so far

Reply:
I've read through the code, thanks for the improvements here with comments and the explanations - very helpful.

I just want to raise https://github.com/dwyl/sendemail/pull/83#pullrequestreview-189591046 as a concern before any further reviewing or work is done in this pr.